staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -149,8 +149,11 @@ func processAuditEvent(sink audit.Sink, ev *auditinternal.Event, omitStages []au
 	} else {
 		ev.StageTimestamp = metav1.NewMicroTime(time.Now())
 	}
+	// Make deep copy of event before sending to the sink, so that as it can be processed concurrently as
+	// the request passes through its stages
+	e := ev.DeepCopy()
```

Comment:
Update the godoc on the ProcessEvents() interface to indicate the event may not be mutated, but is safe to use from spawned goroutines

This change means that when multiple sinks exist (logfile+webhook, logfile+dynamic, etc), the same instance of the audit event will be sent to all of their ProcessEvents handlers. This means we need to make sure that no mutation of the original object is done in *any* ProcessEvents methods. 

TruncateBackend modifies the annotations map in the passed audit event (shallow copy keeps the same annotations map):

https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/truncate/truncate.go#L121-L127

Another concerning implication is what happens when the events are converted to external versions and encoded (along with the nested request/response objects). Do we exercise any of the encoding paths that briefly set group/version/kind?
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L209-L212
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L262-L264
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L231-L233

I see us encoding (or making calls that lead to encoding) in these places:
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go#L74
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/truncate/truncate.go#L141
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go#L96

We need to verify and add tests that exercise those backend paths concurrently to make sure we are not introducing data races with this.

Reply:
Ok @liggitt, circling back on this. As far as the truncate piece goes, the functionality there hasn’t changed from what was existing. That being said it isn’t desirable behavior. I think the simplest solve may be to copy the annotations in the truncate plugin, I've added this to the PR.

For the encoding piece, the shallow copy seems to take care of that, but I’ve added tests to check that we aren’t racing. Let me know what you think

Reply:
when I commented out the shallow copy and annotation copy, the async race tests don't fail. are we exercising the backends we actually use in that test?

Reply:
I'll add another test at the handler, @liggitt would it make sense to feature gate this?